crates/network-scanner-net/src/async_io.rs

```diff
@@ -0,0 +1,150 @@
+                .iter()
+                .map(|waiter| (waiter.id, waiter.waker.clone()))
+                .collect::<std::collections::HashMap<_, _>>();
+            map
+        }
+        let thread_builder = std::thread::Builder::new().name("async-io-event-loop".to_string());
+        let handle = thread_builder.spawn(move || {
+            let mut events = Events::with_capacity(NonZeroUsize::new(1024).unwrap());
+            tracing::trace!("starting io event loop");
+            loop {
+                events.clear();
+
+                tracing::trace!("polling events");
+                poller.wait(&mut events, None).expect("polling failed");
+                let map = get_map(&waiting_list);
+                let mut to_remove = vec![];
+                for event in events.iter() {
+                    tracing::warn!("event {:?}", event);
+                    let key = event.key;
+                    map.get(&key).map(|waker| {
+                        tracing::warn!("waking up waker {:?}", waker);
+                        waker.wake_by_ref();
+                        to_remove.push(key);
+                    });
+                }
+
+                waiting_list
+                    .blocking_lock()
+                    .retain(|waiter| !to_remove.contains(&waiter.id));
+            }
+        })?;
+        self.loop_handle = Some(handle);
+        Ok(self.get_handle())
+    }
+
+    async fn awake_when_ready(&self, socket: Arc<Socket>, event: Event, waker: std::task::Waker) -> anyhow::Result<()> {
+        let waiter = SocketWaiter {
+            id: event.key,
+            waker: waker.clone(),
+        };
+
+        // acquire lock
+        let exist = self
+            .waiting_list
+            .lock()
+            .await
+            .iter()
+            .any(|waiter| waiter.id == event.key);
+
+        if exist {
+            return Ok(());
+        }
+
+        self.waiting_list.lock().await.push(waiter);
+        self.poller.modify(socket.as_ref(), event)?;
+
+        tracing::debug!("awake_when_ready, modified socket {:?} to poller", socket.as_ref());
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AsyncIoHandle {
+    runtime: Arc<Mutex<AsyncIoRuntime>>,
+}
+
+impl AsyncIoHandle {
+    pub async fn new_socket(
+        &mut self,
+        domain: socket2::Domain,
+        ty: socket2::Type,
+        protocol: Option<socket2::Protocol>,
+    ) -> anyhow::Result<TokioRawSocket> {
+        let socket = socket2::Socket::new(domain, ty, protocol)?;
+        let id = self.runtime.lock().await.add_socket(&socket)?;
+        Ok(TokioRawSocket::from_socket(socket, id, self.clone())?)
+    }
+
+    pub(crate) fn awake(&self, socket: Arc<Socket>, id: Event, waker: std::task::Waker) -> anyhow::Result<()> {
+        let runtime = self.runtime.clone();
+        tokio::task::spawn(async move {
+            let runtime = runtime.lock().await;
+            runtime.awake_when_ready(socket, id, waker).await
+        });
```

Comment:
suggestion: Do not use the Mutex from tokio, but the one from parking_lot instead. I suspect we could remove the dependency on `tokio` from the crate entirely.

Reply:
udpated